File: solutions/python/luhn/2/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.digits = card_num.replace(' ','')

    def valid(self):
        digs = self.digits
        if len(digs) < 2 or not digs.isnumeric():
            return False
        ints = list(map(int,digs))
        ints.reverse()
        for i in range(1,len(ints),2):
            ints[i] = ints[i]*2 - 9*(ints[i]>=5)
        return not sum(ints) % 10

File: solutions/ruby/luhn/2/luhn.rb
module Luhn
  def self.valid?(str)
    str = str.delete(' ')
    return false if str.length <= 1
    return false if str.chars.any?{not _1.between?("0","9")}
    digits = str.to_i.digits
    (1..digits.length-1).step(2) do |i|
      digits[i] = digits[i] <= 4 ? digits[i]*2 : digits[i]*2-9
    end
    return digits.sum % 10 == 0
  end
end
